docs/decisions/004_client_side_routing.md

```diff
@@ -0,0 +1,65 @@
+# ADR: Unified client side routing in Juno
+
+## Context
+
+The application requires robust client-side routing with support for:
+
+- Configurable mounting base paths so application router can work with a parent router in the shell app.
+- Optional hashed routing in case shell app don't want to mount child app on certain path or does not have client side routing at all.
+- Data preloading and caching.
+- URL-based state persistence.
+- Support for optional, UI-level state (e.g., modals or panels) without affecting navigation hierarchy.
+
+We evaluated several routing libraries including React Router and TanStack Router. After evaluating feature sets, flexibility, and community support, we decided to adopt **TanStack Router**.
+
+Additionally, to integrate routing with data fetching and ensure URL-driven state management, we considered various combinations of client-side state and URL parsing utilities.
+
+## Decision
+
+We adopted the following architecture for client-side routing and state:
+
+1. **Routing Library**: We use **TanStack Router** to manage client-side routing.
+
+   - The router mounts to a `basePath`, provided via application props, defaulting to `'/'`.
+   - This allows embedding the app under sub-paths without requiring changes to the router configuration.
+
+2. **Hashed Routing Support**:
```

Comment:
It’s worth mentioning in this section that TanStack Router, when using createHashHistory(), copies root-level query parameters into the hash on initialization. This can lead to issues where hash query parameters are misinterpreted or merged incorrectly when appending additional ones like ?key=value... to the URL.

Reply:
I've added your suggestion there, even referenced the issue they've in Tanstack router.